org.eclipse.xtext/src/org/eclipse/xtext/xtext/ecoreInference/EClassifierInfo.java

```diff
@@ -299,14 +306,24 @@ public boolean isFeatureSemanticallyEqualApartFromType(EStructuralFeature f1, ES
 			return result;
 		}
 
-		public boolean isFeatureSemanticallyEqualTo(EStructuralFeature f1, EStructuralFeature f2) {
+		private boolean isFeatureSemanticallyEqualTo(EStructuralFeature f1, EStructuralFeature f2, Set<EClassifier> f2Types) {
+			if (f1 == f2) {
+				return true;
+			}
 			boolean result = isFeatureSemanticallyEqualApartFromType(f1, f2);
 			if (f1 instanceof EReference && f2 instanceof EReference) {
 				EClass f1Type = (EClass) f1.getEType();
```

Comment:
do we need instanceof EClass check here?

Reply:
EReferences will return an EClass from their type. I can add an explicit cast to f1 and use getEReferenceType instead.